File: backend/app/services/posture.py
"""
UMEagleEye - Posture Metrics Service (FR-08-01).
Score calculation and weekly metrics aggregation.
"""

import logging
from typing import Dict, Any, List
from datetime import datetime, timezone
from sqlalchemy import select, func, and_
from sqlalchemy.orm import Session

from app.db.models import Asset, Event, Advisory, PostureMetrics
from app.db.enums import Severity, AdvisoryStatus

logger = logging.getLogger(__name__)


class PostureService:
    """Posture score calculation and metrics aggregation."""

    @staticmethod
    def compute_posture_score(session: Session) -> Dict[str, Any]:
        """Compute the overall security posture score (0-100).

        Formula:
            score = 100 - (critical_penalty + high_penalty + drift_penalty + sla_penalty)
            critical_penalty = open_critical_events × 10 (capped at 40)
            high_penalty     = open_high_events × 5 (capped at 20)
            drift_penalty    = unresolved_drifts × 3 (capped at 20)
            sla_penalty      = sla_breaches × 5 (capped at 20)
        """
        # Count open critical events
        open_critical = session.execute(
            select(func.count(Event.event_id)).where(
                Event.severity == Severity.CRITICAL,
            )
        ).scalar() or 0

        # Count open high events
        open_high = session.execute(
            select(func.count(Event.event_id)).where(
                Event.severity == Severity.HIGH,
            )
        ).scalar() or 0

        # Count unresolved drift events
        drift_types = ["port_opened", "port_closed", "version_downgrade", "config_change", "new_package"]
        unresolved_drifts = session.execute(
            select(func.count(Event.event_id)).where(
                Event.event_type.in_(drift_types),
            )
        ).scalar() or 0

        # Count SLA breaches (Critical advisories still open)
        sla_breaches = session.execute(
            select(func.count(Advisory.advisory_id)).where(
                and_(
                    Advisory.status == AdvisoryStatus.OPEN,
                )
            )
        ).scalar() or 0

        # Calculate penalties
        critical_penalty = min(open_critical * 10, 40)
        high_penalty = min(open_high * 5, 20)
        drift_penalty = min(unresolved_drifts * 3, 20)
        sla_penalty_val = min(sla_breaches * 5, 20)

        score = max(0, min(100, 100 - critical_penalty - high_penalty - drift_penalty - sla_penalty_val))

        # Asset counts
        total_assets = session.execute(select(func.count(Asset.asset_id))).scalar() or 0
        critical_assets = session.execute(
            select(func.count(Asset.asset_id)).where(Asset.criticality_score >= 8)
        ).scalar() or 0

        # Top risks
        top_events = session.execute(
            select(Event).where(
                Event.severity.in_([Severity.CRITICAL, Severity.HIGH])
            ).order_by(Event.composite_risk_score.desc().nullslast()).limit(5)
        ).scalars().all()

        top_risks = [
            {
                "event_type": e.event_type.value,
                "severity": e.severity.value,
                "risk_score": float(e.composite_risk_score) if e.composite_risk_score else 0,
                "details": {k: str(v)[:100] for k, v in (e.details or {}).items()},
            }
            for e in top_events
        ]

        return {
            "overall_score": score,
            "total_assets": total_assets,
            "total_critical_assets": critical_assets,
            "open_critical_events": open_critical + open_high,
            "top_risks": top_risks,
            "breakdown": {
                "critical_penalty": critical_penalty,
                "high_penalty": high_penalty,
                "drift_penalty": drift_penalty,
                "sla_penalty": sla_penalty_val,
            },
        }

    @staticmethod
    def aggregate_weekly_metrics(session: Session) -> Dict[str, Any]:
        """Aggregate weekly data for executive reporting (FR-08-01).

        Returns JSON payload with: total assets, drift events, top CVEs, SLA rate.
        """
        from datetime import timedelta
        week_ago = datetime.now(timezone.utc) - timedelta(days=7)

        total_assets = session.execute(select(func.count(Asset.asset_id))).scalar() or 0

        # Weekly drift events
        drift_count = session.execute(
            select(func.count(Event.event_id)).where(
                Event.timestamp >= week_ago,
            )
        ).scalar() or 0

        # Top CVEs this week
        cve_events = session.execute(
            select(Event).where(
                and_(Event.event_type == "cve_detected", Event.timestamp >= week_ago)
            ).order_by(Event.composite_risk_score.desc().nullslast()).limit(5)
        ).scalars().all()

        top_cves = [
            {
                "cve_id": e.details.get("cve_id", "Unknown"),
                "cvss": e.details.get("cvss_base_score", 0),
                "package": e.details.get("package_name", ""),
            }
            for e in cve_events
        ]

        # SLA resolution rate
        total_advisories = session.execute(
            select(func.count(Advisory.advisory_id)).where(
                Advisory.created_at >= week_ago,
            )
        ).scalar() or 0

        resolved_advisories = session.execute(
            select(func.count(Advisory.advisory_id)).where(
                and_(Advisory.created_at >= week_ago, Advisory.status == AdvisoryStatus.RESOLVED)
            )
        ).scalar() or 0

        sla_rate = (resolved_advisories / total_advisories * 100) if total_advisories > 0 else 100

        return {
            "period": "weekly",
            "total_assets": total_assets,
            "drift_events": drift_count,
            "top_cves": top_cves,
            "sla_resolution_rate": round(sla_rate, 1),
            "total_advisories": total_advisories,
            "resolved_advisories": resolved_advisories,
            "generated_at": datetime.now(timezone.utc).isoformat(),
        }

File: workers/src/services/posture.ts
import type { DB } from '../db/client'
import { assets, events, advisories, postureMetrics } from '../db/schema'
import { eq, and, gte, count, sql } from 'drizzle-orm'

interface PostureResult {
  overallScore: number
  totalAssets: number
  totalCriticalAssets: number
  openCriticalEvents: number
  topRisks: Array<{ asset_id: string; ip: string; hostname: string | null; risk: string }>
}

export async function computePosture(db: DB, tenantId?: string): Promise<PostureResult> {
  // Fetch asset counts
  const assetFilter = tenantId ? eq(assets.tenantId, tenantId) : undefined
  const allAssets = await db
    .select({ assetId: assets.assetId, ipAddress: assets.ipAddress, hostname: assets.hostname,
               criticalityScore: assets.criticalityScore, tenantId: assets.tenantId })
    .from(assets)
    .where(assetFilter)

  const totalAssets = allAssets.length
  const criticalAssets = allAssets.filter(a => (a.criticalityScore ?? 0) >= 8)
  const totalCriticalAssets = criticalAssets.length

  // Open critical events in last 7 days
  const sevenDaysAgo = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000)
  const criticalEventRows = await db
    .select({ eventId: events.eventId, assetId: events.assetId })
    .from(events)
    .where(and(
      eq(events.severity, 'critical'),
      gte(events.timestamp, sevenDaysAgo)
    ))

  // Filter to this tenant's assets
  const tenantAssetIds = new Set(allAssets.map(a => a.assetId))
  const openCritical = criticalEventRows.filter(e => tenantAssetIds.has(e.assetId))
  const openCriticalEvents = openCritical.length

  // High events in last 7 days
  const highEventRows = await db
    .select({ eventId: events.eventId })
    .from(events)
    .where(and(eq(events.severity, 'high'), gte(events.timestamp, sevenDaysAgo)))

  const openHighEvents = highEventRows.filter(e =>
    'assetId' in e ? tenantAssetIds.has((e as { assetId: string }).assetId) : false
  ).length

  // Score calculation
  let score = 100
  score -= Math.min(openCriticalEvents * 5, 40)   // -5 per critical, max -40
  score -= Math.min(openHighEvents * 2, 20)         // -2 per high, max -20
  if (totalAssets > 0 && totalCriticalAssets / totalAssets > 0.2) {
    score -= 10  // >20% assets are high-criticality
  }
  score = Math.max(0, Math.min(100, score))

  // Top risks: assets with most recent critical events
  const topRisks = criticalAssets.slice(0, 5).map(a => ({
    asset_id: a.assetId,
    ip: a.ipAddress,
    hostname: a.hostname ?? null,
    risk: 'High criticality asset',
  }))

  return {
    overallScore: Math.round(score),
    totalAssets,
    totalCriticalAssets,
    openCriticalEvents,
    topRisks,
  }
}

export async function savePostureSnapshot(db: DB, tenantId?: string): Promise<void> {
  const result = await computePosture(db, tenantId)
  await db.insert(postureMetrics).values({
    tenantId: tenantId ?? null,
    overallScore: result.overallScore,
    totalAssets: result.totalAssets,
    totalCriticalAssets: result.totalCriticalAssets,
    openCriticalEvents: result.openCriticalEvents,
    topRisks: result.topRisks,
  })
}
